### PR TITLE
roachtest: fix tenant auto upgrade scheduling user hooks during setup

### LINKED ISSUE #145951: roachtest: cdc/mixed-version/checkpointing failed
roachtest.cdc/mixed-version/checkpointing [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/19595977?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/19595977?buildTab=artifacts#/cdc/mixed-version/checkpointing) on master @ [6b5994cb063e494fa7013c731170e5f539d9ba96](https://github.com/cockroachdb/cockroach/commits/6b5994cb063e494fa7013c731170e5f539d9ba96). A Side-Eye cluster snapshot was captured on timeout: [https://app.side-eye.io/#/snapshots/807](https://app.side-eye.io/#/snapshots/807).

```
(test_runner.go:1417).runTest: test timed out (3h0m0s)
test artifacts and logs in: /artifacts/cdc/mixed-version/checkpointing/run_1
```

Parameters:
 - <code>arch=amd64</code>
 - <code>cloud=gce</code>
 - <code>coverageBuild=false</code>
 - <code>cpu=4</code>
 - <code>encrypted=false</code>
 - <code>fs=ext4</code>
 - <code>localSSD=true</code>
 - <code>metamorphicBufferedSender=false</code>
 - <code>mvtDeploymentMode=shared-process</code>
 - <code>mvtVersions=v24.2.9 → v24.3.9 → v25.1.0 → master</code>
 - <code>runtimeAssertionsBuild=false</code>
 - <code>ssd=0</code>
<details><summary>Help</summary>
<p>

See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)

See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

See: [Grafana](https://go.crdb.dev/roachtest-grafana/teamcity-19595977/cdc-mixed-version-checkpointing/1746171933888/1746182881341)

</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*cdc/mixed-version/checkpointing.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

Jira issue: CRDB-50299

### PR DESCRIPTION
**Pull request by DarrylWong**
The mixed version planner splits up a plan into two partitions, setup steps and test steps. The latter will have user hooks scheduled while the former will omit them.

The scheduling of user hooks is gated by the scheduleHooks boolean. However, there was one edge case which omitted this check and allowed for user hooks to be scheduled during version finalization of a tenant.

Fixes: #145951
Release note: none

### CODE REVIEW

**Review by herkolategan — APPROVED** [review]

File: pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go [hunk 1]
```diff
@@ -938,10 +938,12 @@ func (p *testPlanner) finalizeUpgradeSteps(
 		// auto-upgrading after resetting a cluster setting, we need to
 		// "manually" run the migrations at this point.
 		if toVersion.AtLeast(tenantSupportsAutoUpgradeVersion) {
-			steps = append(
-				steps,
-				p.concurrently(mixedVersionLabel, p.hooks.MixedVersionSteps(p.currentContext, p.prng))...,
-			)
+			if scheduleHooks {
```
Comment (srosenberg) [inline pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go:941]: To facilitate future debugging of this type of issue, we should probably update `MixedVersionSteps`, `AfterUpgradeFinalizedSteps`, etc., with the enclosing context. E.g., instead of the ambiguous,

```
run "wait and validate" (16)
```

we could say something like,

```
mvt.InMixedVersion--"wait and validate" (16)
```

to denote the specific type of user-hook.

**Review by srosenberg — APPROVED** [review]
Thanks!

File: pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go (hunk 1 above)
Comment (DarrylWong) [reply pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go:941]: Theres actually a debug mode used by the unit tests that will print the stage of each step. Let me look into it sometime this week, sounds like exactly what we want. I'll merge this PR for now to stop any flakes.

**Comment by DarrylWong** [comment]
TFTR!

bors r=herkolategan, srosenberg

File: pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go (hunk 1 above)
Comment (srosenberg) [reply pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go:941]: > Theres actually a debug mode used by the unit tests that will print the stage of each step.

With `prettyPrintInternal(true)`, it prints the enclosing stage (`ss.context.System.Stage`), so it may look like,

```
run "wait and validate" (16) [stage=running-upgrade-migrations,finalizing]
```

but it's still _not_ immediately obvious what user-hook is being run.

**Comment by DarrylWong** [comment]
Unrelated test flake. Coincidentally a mixed version test but from the logic test framework.

bors retry
